### PR TITLE
Backward compatable license sigining

### PR DESCRIPTION
**Pull request by salahalsaleh1993**

### CODE REVIEW

File: pkg/pull/signature.go [hunk 1]
```diff
@@ -17,29 +17,49 @@ var (
...
+	LicenseData    string `json:"licenseData"`
+	InnerSignature string `json:"innerSignature"`
+}
+
 type KeySignature struct {
 	Signature   []byte `json:"signature"`
 	GlobalKeyId string `json:"globalKeyId"`
 }
 
 func VerifySignature(license *kotsv1beta1.License) error {
```
Comment (laverya) [inline pkg/pull/signature.go]: this should return the license data from the internal base64 string

File: kotskinds/apis/kots/v1beta1/license_types.go [hunk 2]
```diff
@@ -99,7 +99,7 @@ type EntitlementField struct {
 
 // LicenseSpec defines the desired state of LicenseSpec
 type LicenseSpec struct {
-	Signature         []byte                      `json:"signature"`
+	Signature         string                      `json:"signature"`
```
Comment (divolgin) [inline kotskinds/apis/kots/v1beta1/license_types.go]: if you keep these as []byte, you won't have to do `base64.StdEncoding.DecodeString(license.Spec.Signature)`

File: pkg/pull/signature.go [hunk 3]
```diff
@@ -48,16 +68,16 @@ func VerifySignature(license *kotsv1beta1.License) error {
...
 
-	licenseMessage, err := getMessageFromLicense(license)
+	licenseMessage, err := base64.StdEncoding.DecodeString(outerSignature.LicenseData)
 	if err != nil {
-		return errors.Wrap(err, "failed to convert license to message")
+		return errors.New("failed to decode license data from outer signature")
 	}
 
-	if err := verify(licenseMessage, signature.LicenseSignature, []byte(signature.PublicKey)); err != nil {
+	if err := verify(licenseMessage, innerSignature.LicenseSignature, []byte(innerSignature.PublicKey)); err != nil {
```
Comment (laverya) [inline pkg/pull/signature.go]: this verify can be run against the base64 encoded data itself
but really what matters is doing it the same way everywhere, I suppose
```suggestion
	if err := verify(outerSignature.LicenseData, innerSignature.LicenseSignature, []byte(innerSignature.PublicKey)); err != nil {
```

File: kotskinds/apis/kots/v1beta1/license_types.go (hunk 2 above)
Comment (salahalsaleh1993) [reply kotskinds/apis/kots/v1beta1/license_types.go]: noted, will update

Comment (laverya) [reply kotskinds/apis/kots/v1beta1/license_types.go]: You'd still need a base64 decode of some form though?

File: pkg/pull/signature.go (hunk 3 above)
Comment (salahalsaleh1993) [reply pkg/pull/signature.go]: we sign the data before encoding it, we can update both if that's the preferred approach

File: pkg/pull/signature.go [hunk 4]
```diff
@@ -48,16 +68,16 @@ func VerifySignature(license *kotsv1beta1.License) error {
...
 		return errors.Wrap(err, "failed to verify license signature")
 	}
```
Comment (laverya) [inline pkg/pull/signature.go]: before returning here you should parse `licenseMessage` as a LicenseSpec
and then compare it to the other fields within the passed license's LicenseSpec (everything besides 'signature')

File: pkg/pull/signature.go (hunk 1 above)
Comment (laverya) [reply pkg/pull/signature.go]: (or maybe overwrite it, since pass-by-reference?)

File: kotskinds/apis/kots/v1beta1/license_types.go (hunk 2 above)
Comment (salahalsaleh1993) [reply kotskinds/apis/kots/v1beta1/license_types.go]: apparently not

File: pkg/pull/signature.go (hunk 3 above)
Comment (laverya) [reply pkg/pull/signature.go]: If it's easier to sign before encoding, we can do that
(I don't have a preference here)

File: pkg/pull/signature.go [hunk 5]
```diff
@@ -88,41 +101,48 @@ func verify(message, signature, publicKeyPEM []byte) error {
...
-		valStr, err := json.Marshal(val)
-		if err != nil {
-			return nil, errors.Wrapf(err, "failed to marshal entitlement value: %s", k)
-		}
-		fields[key] = string(valStr)
+func verifyLicenseData(outerLicense *kotsv1beta1.License, innerLicense *kotsv1beta1.License) error {
+	if outerLicense.Spec.AppSlug != innerLicense.Spec.AppSlug {
+		return errors.New("\"appSlug\" field has changed")
+	}
+	if outerLicense.Spec.Endpoint != innerLicense.Spec.Endpoint {
```
Comment (emosbaugh) [inline pkg/pull/signature.go]: @marccampbell mentioned that he may want to be able to change the endpoint. you should ask him about this

Comment (laverya) [reply pkg/pull/signature.go]: perhaps that should be done with an override param elsewhere?
And also have "can override endpoint" be a license param

File: pkg/pull/signature.go [hunk 6]
```diff
@@ -17,51 +16,65 @@ var (
...
-	if err != nil {
-		return errors.Wrap(err, "failed to convert license to message")
+	if err := verify(outerSignature.LicenseData, innerSignature.LicenseSignature, []byte(innerSignature.PublicKey)); err != nil {
+		return nil, errors.Wrap(err, "failed to verify license signature")
 	}
 
-	if err := verify(licenseMessage, signature.LicenseSignature, []byte(signature.PublicKey)); err != nil {
-		return errors.Wrap(err, "failed to verify license signature")
+	verifiedLicense := &kotsv1beta1.License{}
+	if err := json.Unmarshal(outerSignature.LicenseData, verifiedLicense); err != nil {
```
Comment (laverya) [inline pkg/pull/signature.go]: I believe your current PR has `outerSignature.LicenseData` being base64'd here
One of the two will need to change (IMO `outerSignature.LicenseData` doesn't need base64, so you should remove it in vandoor)
Unless I'm misreading things

Comment (divolgin) [reply pkg/pull/signature.go]: Here it's defined as `[]byte` so it's automatically decoded.

Comment (salahalsaleh1993) [reply pkg/pull/signature.go]: Updated in license-api

File: pkg/pull/signature.go (hunk 5 above)
Comment (marccampbell) [reply pkg/pull/signature.go]: @salahalsaleh1993 I agree with Ethan. The endpoint parameter should be configurable by the end customer here. It’s not part of the product yet, but some end customers could have proxies and want to redirect the endpoint to a proxy or other policy engine to pull. 

Let’s keep endpoint in the license spec, but it should not be validated during the signature.

Comment (salahalsaleh1993) [reply pkg/pull/signature.go]: cool, will update, thanks!

**Pull request merged**